util/generate_compilation_db.py

```diff
@@ -0,0 +1,150 @@
+jump-to-definition and cross-references in IDEs that support
+compile_commands.json.
+
+The analysis.ActionGraphContainer protobuf [0] defines aquery's results format.
+Clang informally defines the schema of compile_commands.json [1].
+
+Caveat: this tool only emits the commands for building C/C++ code.
+
+Example:
+  util/generate_compilation_db.py --target //sw/... --out compile_commands.json
+
+[0]: https://github.com/bazelbuild/bazel/blob/master/src/main/protobuf/analysis_v2.proto
+[1]: https://clang.llvm.org/docs/JSONCompilationDatabase.html
+
+"""
+
+import argparse
+import json
+import os
+import subprocess
+
+
+def build_id_lookup_dict(dicts: list[dict]):
+    """Create a dict from `dicts` indexed by the "id" key."""
+    lookup = {}
+    for d in dicts:
+        lookup[d['id']] = d
+    return lookup
+
+
+class BazelAqueryResults:
+    """Corresponds to Bazel's analysis.ActionGraphContainer protobuf."""
+    def __init__(self, output: str):
+        parsed = json.loads(output)
+        self.actions = [BazelAqueryAction(action) for action in parsed['actions']]
+        self.dep_sets_ = build_id_lookup_dict(parsed['depSetOfFiles'])
+        self.artifacts_ = build_id_lookup_dict(parsed['artifacts'])
+        self.path_fragments_ = build_id_lookup_dict(parsed['pathFragments'])
+
+    def reconstruct_path(self, id: int):
+        """Reconstruct a file path from Bazel aquery fragments."""
+        labels = []
+
+        while True:
+            path_fragment = self.path_fragments_[id]
+            labels.append(path_fragment['label'])
+
+            if 'parentId' not in path_fragment:
+                break
+            id = path_fragment['parentId']
+
+        return os.path.join(*reversed(labels))
+
+    def iter_artifacts_for_dep_sets(self, dep_set_ids: list[int]):
+        """Iterate the reconstructed paths of all artifacts related to `dep_set_ids`."""
+
+        dep_set_id_stack = dep_set_ids
+        while len(dep_set_id_stack) > 0:
+            dep_set_id = dep_set_id_stack.pop()
+            dep_set = self.dep_sets_[dep_set_id]
+
+            for direct_artifact_id in dep_set.get('directArtifactIds', []):
+                artifact = self.artifacts_[direct_artifact_id]
+                path_fragment_id = artifact['pathFragmentId']
+                path = self.reconstruct_path(path_fragment_id)
+                yield path
+
+            for transitive_dep_set_id in dep_set.get('transitiveDepSetIds', []):
+                dep_set_id_stack.append(transitive_dep_set_id)
+
+
+class BazelAqueryAction:
+    """Corresponds to Bazel's analysis.Action protobuf."""
+    def __init__(self, action: dict):
+        self.mnemonic = action.get('mnemonic', None)
+        self.arguments = action.get('arguments', None)
+        self.input_dep_set_ids = action.get('inputDepSetIds', [])
+
+    def hackily_hardcode_googletest_includes(self):
+        """Add the include paths for googletest (gtest and gmock) to `self.arguments`."""
```

Comment:
Could we do without this if we properly added googletest to our bazel config rather than it's current addition via a `local_repository` rule?

Reply:
Possibly! But I'm still not clear why the extracted build commands are not sufficient to find these headers.

This must be the `local_repository` rule you're referring to[^0]. How are you thinking we'd add googletest instead?

The quickstart instructions for googletest[^1] recommends using the `http_archive` rule, but the Bazel docs[^2] seem ambivalent on whether you use `local_repository`, `git_repository`, or `http_archive` for external deps.

[^0]: https://cs.opensource.google/opentitan/opentitan/+/master:third_party/cc/repos.bzl;l=23;drc=25be34391987724e412c8bd6efcfd81cc03d1212
[^1]: https://google.github.io/googletest/quickstart-bazel.html
[^2]: https://bazel.build/docs/external